framework/src/base/MooseApp.C

```diff
@@ -358,10 +364,22 @@ MooseApp::setInputFileName(std::string input_filename)
   _input_filename = input_filename;
 }
 
+const MooseApp &
+MooseApp::root() const
```

Comment:
We shouldn't debate the merits/drawbacks of switching the Multiapps to parent pointers instead of child pointers in this PR. Either design is valid for a tree based structure. I think you both have good points but we don't have to change what's working right here, right now. I can see that this PR does remove several lines of code but we could maintain the child pointer design *and* pass everything in to the newly constructed apps so that these methods could still be removed. Everything else in this PR looks reasonable.